README.rst
============
``gs.dmarc``
============
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Look up and report on the DMARC status of a domain
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

:Author: `Michael JasonSmith`_
:Contact: Michael JasonSmith <mpj17@onlinegroups.net>
:Date: 2015-06-25
:Organization: `GroupServer.org`_
:Copyright: This document is licensed under a
  `Creative Commons Attribution-Share Alike 4.0 International License`_
  by `OnlineGroups.net`_.

..  _Creative Commons Attribution-Share Alike 4.0 International License:
    http://creativecommons.org/licenses/by-sa/4.0/

Introduction
============

This product allows systems look up and report on the DMARC (`RFC
7489`_: Domain-based Message Authentication, Reporting and
Conformance) status of a domain. DMARC allows the owner of a
domain to publish a key that is used to verify if an email
message actually originated from the domain, and to publish what
to do if the verification fails. It is an extension of DKIM (`RFC
6376`_: DomainKeys Identified Mail) and SPF (`RFC 4408`_: Sender
Policy Framework).

Specifically this product supplies ``gs.dmarc.ReceiverPolicy``
for enumerating [#enum34]_ the different DMARC policies, and the
``receiver_policy`` function for querying the policy for a
given domain.

Resources
=========

- Documentation: http://gsdmarc.readthedocs.io/
- Code repository: https://github.com/groupserver/gs.dmarc
- Questions and comments to
  http://groupserver.org/groups/development
- Report bugs at https://redmine.iopen.net/projects/groupserver

.. _RFC 7489: https://tools.ietf.org/html/rfc7489.html
.. _RFC 6376: http://tools.ietf.org/html/rfc6376
.. _RFC 4408: http://tools.ietf.org/html/rfc4408
.. [#enum34] `The enum34 package`_ is used to provide `Enum`_
           support for releases of Python prior to 3.4.
.. _The enum34 package: https://pypi.python.org//pypi/enum34
.. _Enum: https://docs.python.org/3/library/enum.html
.. _GroupServer: http://groupserver.org/
.. _GroupServer.org: http://groupserver.org/
.. _OnlineGroups.Net: https://onlinegroups.net
.. _Michael JasonSmith: http://groupserver.org/p/mpj17

..  LocalWords:  DMARC DKIM DomainKeys dkim groupserver spf enum
..  LocalWords:  lookup


docs/HISTORY.rst
Changelog
=========

2.1.9 (2016-12-16)
------------------

* Falling back to the either the overall subdomain-policy or
  overall domain-policy if a subdomain lacks a specific published
  DMARC record — with thanks to Igor Colombi for pointing out
  the issue

2.1.8 (2016-10-18)
------------------

* Adding :pep:`484` type hints
* Updating the public-suffix list
* Using :mod:`setuptools` to return the public-suffix list

2.1.7 (2016-04-11)
------------------

* Testing with Python 3.5
* Switching to dictionary-comprehensions

2.1.6 (2016-03-24)
------------------

* Updating the suffix list from Mozilla, thanks to `Baran
  Kaynak`_

.. _Baran Kaynak: https://github.com/barankaynak

2.1.5 (2015-09-01)
------------------

* Catching ``dns.resolver.NoNameserver`` exceptions, thanks to
  `Alexy Mikhailichenko`_

.. _Alexy Mikhailichenko: https://github.com/alexymik

2.1.4 (2015-06-25)
------------------

* Fixing a spelling mistake in the README, thanks to `Stefano
  Brentegani`_
* Updating the documentation, as DMARC is now :rfc:`7489`

.. _Stefano Brentegani: https://github.com/brente

2.1.3 (2014-10-20)
------------------

* Handling domains with invalid DMARC policies, closing `Bug 4135`_

.. _Bug 4135: <https://redmine.iopen.net/issues/4135

2.1.2 (2014-09-26)
------------------

* Switching to GitHub_ as the primary code repository.

.. _GitHub: https://github.com/groupserver/gs.dmarc

2.1.1 (2014-07-09)
------------------

* Coping when the host-name passed to ``lookup_receiver_policy``
  for hosts that start with ``_dmarc`` already
* Rejecting all answers that do not start with ``v=DMARC1``, as
  per `Section 7.1`_ (number 5) of the draft DMARC specification

.. _Section 7.1: https://tools.ietf.org/html/rfc7489#section-7.1

2.1.0 (2014-05-07)
------------------

* Adding ``gs.dmarc.receiver_policy``, which looks up the
  organisational domain
* Updating the Sphinx documentation

2.0.0 (2014-04-29)
------------------

* Adding ``gs.dmarc.ReceiverPolicy.noDmarc``, and returning it
  from ``gs.dmarc.lookup_receiver_policy``
* Adding Sphinx documentation

1.0.0 (2014-04-24)
------------------

Initial release.

..  LocalWords:  Changelog GitHub README


gs/dmarc/tests/lookup.py
# -*- coding: utf-8 -*-
############################################################################
#
# Copyright © 2014, 2015, 2016 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
############################################################################
from __future__ import absolute_import, unicode_literals, print_function
from unittest import TestCase
import dns.resolver
from mock import patch
import gs.dmarc.lookup


class TestLookup(TestCase):
    '''Test the lookup of DMARC policies'''

    @staticmethod
    def create_response(policy):
        'Create a fake DMARC response, with the specified ``policy``'
        r = '"v=DMARC1; p={0}; sp=none; pct=100; '\
            'rua=mailto:dmarc-example-rua@example.com, '\
            'mailto:dmarc_e_rua@example.com;"'
        retval = [r.format(policy)]
        return retval

    def lookup_receiver_policy(self, policy):
        'Perform a fake lookup of the DMARC reciever policy'
        queryResp = self.create_response(policy)
        with patch('gs.dmarc.lookup.dns_query') as faux_query:
            faux_query.return_value = queryResp
            retval = gs.dmarc.lookup.lookup_receiver_policy('example.com')
        return retval

    def assertPolicy(self, expected, val):
        m = 'Expected the policy {0}, got {1}'
        msg = m.format(expected, val)

        self.assertEqual(expected, val, msg)

    def test_lookup_none(self):
        'Test the reciever-policy "none"'
        r = self.lookup_receiver_policy('none')

        self.assertPolicy(gs.dmarc.lookup.ReceiverPolicy.none, r)

    def test_lookup_reject(self):
        'Test the reciever-policy "reject"'
        r = self.lookup_receiver_policy('reject')

        self.assertPolicy(gs.dmarc.lookup.ReceiverPolicy.reject, r)

    def test_lookup_quarantine(self):
        'Test the reciever policy "quarantine".'
        r = self.lookup_receiver_policy('quarantine')

        self.assertPolicy(gs.dmarc.lookup.ReceiverPolicy.quarantine, r)

    def test_lookup_monitor(self):
        'Test the reciever-policy "monitor", which is not a reciever policy'
        r = self.lookup_receiver_policy('monitor')

        self.assertPolicy(gs.dmarc.lookup.ReceiverPolicy.noDmarc, r)

    @patch('gs.dmarc.lookup.dns_query')
    def test_lookup_nxdomain(self, faux_query):
        'Test a failed lookup of a domain (non-existant domain).'
        faux_query.side_effect = dns.resolver.NXDOMAIN
        r = gs.dmarc.lookup.lookup_receiver_policy('example.com')

        self.assertPolicy(gs.dmarc.lookup.ReceiverPolicy.noDmarc, r)

    @patch('gs.dmarc.lookup.dns_query')
    def test_lookup_noanswer(self, faux_query):
        'Test a failed lookup of a domain (no answer).'
        faux_query.side_effect = dns.resolver.NoAnswer
        r = gs.dmarc.lookup.lookup_receiver_policy('example.com')

        self.assertPolicy(gs.dmarc.lookup.ReceiverPolicy.noDmarc, r)

    @patch('gs.dmarc.lookup.dns_query')
    def test_lookup_nonameservers(self, faux_query):
        'Test a failed lookup of a domain (no name servers).'
        faux_query.side_effect = dns.resolver.NoNameservers
        r = gs.dmarc.lookup.lookup_receiver_policy('example.com')

        self.assertPolicy(gs.dmarc.lookup.ReceiverPolicy.noDmarc, r)

    @patch('gs.dmarc.lookup.dns_query')
    def test_lookup_malformed(self, faux_query):
        '''Test that a response that does not start with "v= is interpreted
        as no DMARC.'''
        # A 'q' is not a '"'.
        a = self.create_response('reject')
        answer = 'q' + a[0]
        queryResp = [answer]
        faux_query.return_value = queryResp
        r = gs.dmarc.lookup.lookup_receiver_policy('example.com')

        self.assertPolicy(gs.dmarc.lookup.ReceiverPolicy.noDmarc, r)

    def test_lookup_invalid_policy(self):
        '''Test that an invalid policy-tag raises an error'''
        with self.assertRaises(ValueError):
            gs.dmarc.lookup_receiver_policy('example.com', policyTag='Piranha')

    @patch('gs.dmarc.lookup.dns_query')
    def test_lookup_policy_tag(self, faux_query):
        '''Test that explicitly using the ``p``policy tag is fine.'''
        queryResp = self.create_response('reject')
        faux_query.return_value = queryResp
        r  = gs.dmarc.lookup.lookup_receiver_policy('example.com', policyTag='p')

        self.assertPolicy(gs.dmarc.lookup.ReceiverPolicy.reject, r)

    @patch('gs.dmarc.lookup.dns_query')
    def test_lookup_subdomain_policy_tag(self, faux_query):
        '''Test that explicitly using the ``sp``policy tag is fine.'''
        queryResp = self.create_response('reject')
        faux_query.return_value = queryResp
        r  = gs.dmarc.lookup.lookup_receiver_policy('example.com', policyTag='sp')

        self.assertPolicy(gs.dmarc.lookup.ReceiverPolicy.none, r)

    @patch('gs.dmarc.lookup.dns_query')
    def test_lookup_policy_fallback(self, faux_query):
        '''Test that the value of the ``p`` tag is returned if the ``sp``policy is absent.'''
        q = self.create_response('reject')[0]
        # Remove the subdomain policy from the DNS return
        faux_query.return_value = [q.replace('sp=none; ', ''), ]
        r  = gs.dmarc.lookup.lookup_receiver_policy('example.com', policyTag='sp')

        self.assertPolicy(gs.dmarc.lookup.ReceiverPolicy.reject, r)


gs/dmarc/tests/__init__.py
# -*- coding: utf-8 -*-
# This space deliberately left blank


gs/dmarc/tests/receiver.py
# -*- coding: utf-8 -*-
############################################################################
#
# Copyright © 2014 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
############################################################################
from __future__ import absolute_import, unicode_literals
from unittest import TestCase
from mock import MagicMock
import gs.dmarc.lookup
from gs.dmarc.lookup import (ReceiverPolicy, receiver_policy,
                             PublicSuffixList as psl)


class TestReceiverPolicy(TestCase):
    '''Test the receiver policy'''

    def test_suffix_file(self):
        'Test that we can read the data file for the public suffix list'
        fn = gs.dmarc.lookup.get_suffix_list_file_name()
        self.assertEqual('txt', fn[-3:])
        with open(fn, 'r') as infile:
            l = infile.readline()
        self.assertEqual('// This Source Code Form', l[:24])

    def test_receiver_policy_none(self):
        'Test that we only look up the reciever policy once if it is "none"'
        gs.dmarc.lookup.lookup_receiver_policy = \
            MagicMock(return_value=ReceiverPolicy.none)
        psl.get_public_suffix = MagicMock(return_value='example.com')
        r = receiver_policy('example.com')

        cc = gs.dmarc.lookup.lookup_receiver_policy.call_count
        self.assertEqual(1, cc)
        self.assertEqual(0, psl.get_public_suffix.call_count)
        self.assertEqual(r, ReceiverPolicy.none)

    def test_reciver_policy_noDmarc(self):
        'Test that we look up the public suffix for the host for "no dmarc"'
        gs.dmarc.lookup.lookup_receiver_policy = \
            MagicMock(return_value=ReceiverPolicy.noDmarc)
        psl.get_public_suffix = MagicMock(return_value='example.com')
        r = receiver_policy('my.example.com')

        cc = gs.dmarc.lookup.lookup_receiver_policy.call_count
        self.assertEqual(2, cc)
        self.assertEqual(gs.dmarc.lookup.lookup_receiver_policy.call_args,
                         (('example.com', ), {'policyTag': 'sp'}),
                         'Failed to look up the subdomain policy')
        self.assertEqual(1, psl.get_public_suffix.call_count)
        self.assertEqual(r, ReceiverPolicy.noDmarc)

    def test_dmarc_host(self):
        'Test a host lookup when "_dmarc." something is given'
        gs.dmarc.lookup.lookup_receiver_policy = \
            MagicMock(return_value=ReceiverPolicy.none)
        psl.get_public_suffix = MagicMock(return_value='example.com')
        r = receiver_policy('_dmarc.example.com')

        cc = gs.dmarc.lookup.lookup_receiver_policy.call_count
        self.assertEqual(1, cc)
        self.assertEqual(0, psl.get_public_suffix.call_count)
        self.assertEqual(r, ReceiverPolicy.none)

        args, kw_args = gs.dmarc.lookup.lookup_receiver_policy.call_args
        self.assertEqual('example.com', args[0])


version.py
# -*- coding: utf-8 -*-
version = '2.1.9'
release = True

#--------------------------------------------------------------------------#
import sys
if (sys.version_info < (3, )):
    from commands import getstatusoutput
else:
    from subprocess import getstatusoutput  # lint:ok
import datetime
import os
import glob


class CommandError(Exception):
    pass


def execute_command(commandstring):
    status, output = getstatusoutput(commandstring)
    if status != 0:
        m = 'Command "{0}" exited with status {1}'
        msg = m.format(commandstring, status)
        raise CommandError(msg)
    return output


def parse_version_from_package():
    try:
        pkginfo = os.path.join(glob.glob('*.egg-info')[0],
                               'PKG-INFO')
    except:
        pkginfo = ''

    version_string = ''
    if os.path.exists(pkginfo):
        for line in open(pkginfo):
            if line.find('Version: ') == 0:
                version_string = line.strip().split('Version: ')[1].strip()
        if not version_string:
            version_string = '%s-dev' % version
    else:
        version_string = version

    return version_string


def get_version():
    try:
        globalid = execute_command("hg identify -i").strip('+')
        c = "hg log -r %s --template '{date|isodatesec}'" % globalid
        commitdate = execute_command(c)
        # convert date to UTC unix timestamp, using the date command because
        # python date libraries do not stabilise till about 2.6
        dateCommand = 'date -d"%s" --utc +%%s' % commitdate
        timestamp = int(execute_command(dateCommand))

        # finally we have something we can use!
        dt = datetime.datetime.utcfromtimestamp(timestamp)
        datestring = dt.strftime('%Y%m%d%H%M%S')
        if release:
            version_string = version
        else:
            version_string = "%s.dev%s" % (version, datestring, globalid)

    except (CommandError, ValueError, TypeError):
        # --=mpj17=-- Usually because we are building out a source-egg,
        # rather than from a Hg source-directory.
        version_string = parse_version_from_package()

    return version_string

if __name__ == '__main__':
    import sys
    sys.stdout.write('{0}\n'.format(get_version()))


gs/dmarc/lookup.py
# -*- coding: utf-8 -*-
############################################################################
#
# Copyright © 2014, 2015, 2016 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
############################################################################
from __future__ import absolute_import, unicode_literals, print_function
from enum import Enum
from pkg_resources import resource_filename  # Part of setuptools
try:
    # typing is needed by mypy, but is unused otherwise
    from typing import Dict, Text  # noqa: F401
except ImportError:
    pass
from dns.resolver import (query as dns_query, NXDOMAIN, NoAnswer,
                          NoNameservers)
from publicsuffix import PublicSuffixList


class ReceiverPolicy(Enum):
    '''An enumeration of the different receiver policies in DMARC.'''
    __order__ = 'noDmarc none quarantine reject'  # only needed in 2.x

    #: No published DMARC receiver-policy could be found. Often interpreted
    #: the same way as :attr:`gs.dmarc.ReceiverPolicy.none`.
    noDmarc = 0

    #: The published policy is ``none``. Recieving parties are supposed to
    #: skip the verification of the DMARC signature.
    none = 1

    #: Generally causes the message to be marked as *spam* if verification
    #: fails.
    quarantine = 2

    #: Causes the system that is receiving the message to reject the
    #: message if the verification fails.
    reject = 3


def answer_to_dict(answer):
    # type: (Text) -> Dict[unicode, unicode]
    '''Turn the DNS DMARC answer into a dict of tag:value pairs.'''
    a = answer.strip('"').strip(' ')
    rawTags = [t.split('=') for t in a.split(';') if t]
    retval = {t[0].strip(): t[1].strip() for t in rawTags}
    return retval


def lookup_receiver_policy(host, policyTag='p'):
    # type: (str, str) -> ReceiverPolicy
    '''Lookup the reciever policy for a host. Returns a ReceiverPolicy.

:param str host: The host to query. The *actual* host that is queried has
                 ``_dmarc.`` prepended to it.
:param str policyTag: The *tag* that holds the receiver policy. Must be
                      ``p`` (the default) or ``sp`` (for the subdomain
                      policy). See :rfc:`7489#section-6.3`.
:returns: The DMARC receiver policy for the host. If there is no published
          policy then :attr:`gs.dmarc.ReceiverPolicy.noDmarc` is returned.
:rtype: A member of the :class:`gs.dmarc.ReceiverPolicy` enumeration.'''
    if policyTag not in ('p', 'sp'):
        raise ValueError('policyTag must be "p" or "sp".')
    dmarcHost = '_dmarc.{0}'.format(host)
    retval = ReceiverPolicy.noDmarc
    try:
        dnsAnswer = dns_query(dmarcHost, 'TXT')
    except (NXDOMAIN, NoAnswer, NoNameservers):
        pass  # retval = ReceiverPolicy.noDmarc
    else:
        answer = str(dnsAnswer[0])
        # Check that v= field is the first one in the answer (which is in
        # double quotes) as per Section 7.1 (5):
        #     In particular, the "v=DMARC1" tag is mandatory and MUST appear
        #     first in the list. Discard any that do not pass this test.
        # http://tools.ietf.org/html/draft-kucherawy-dmarc-base-04#section-7.1
        if answer[:9] == '"v=DMARC1':
            tags = answer_to_dict(answer)
            # TODO: check that 'none' is the right assumption?
            if policyTag == 'p':
                p = tags.get('p', 'none')
            else:
                # Look up the subdomain policy, falling back to the main policy
                # <https://tools.ietf.org/html/rfc7489#section-6.3>
                p = tags.get('sp', tags.get('p', 'none'))
            policy = p if hasattr(ReceiverPolicy, p) else 'noDmarc'
            # https://github.com/python/mypy/issues/1381
            retval = ReceiverPolicy[policy]  # type: ignore
    return retval


def receiver_policy(host):
    # type: (str) -> ReceiverPolicy
    '''Get the DMARC receiver policy for a host.

:param str host: The host to lookup.
:returns: The DMARC reciever policy for the host.
:rtype:  A member of the :class:`gs.dmarc.ReceiverPolicy` enumeration.

The :func:`receiver_policy` function looks up the DMARC reciever polciy
for ``host``. If the host does not have a pubished policy
`the organizational domain`_ is determined. The DMARC policy for the
organizational domain is queried, and the subdomain policy is reuturned
(if specified) or the overall policy for the domain is returned.
Internally the :func:`gs.dmarc.lookup.lookup_receiver_policy` is used to
perform the query.

.. _the organizational domain:
   http://tools.ietf.org/html/draft-kucherawy-dmarc-base-04#section-3.2'''
    hostSansDmarc = host if host[:7] != '_dmarc.' else host[7:]

    retval = lookup_receiver_policy(hostSansDmarc)
    if retval == ReceiverPolicy.noDmarc:
        fn = get_suffix_list_file_name()
        with open(fn) as suffixList:
            psl = PublicSuffixList(suffixList)
            newHost = psl.get_public_suffix(hostSansDmarc)
        retval = lookup_receiver_policy(newHost, policyTag='sp')
    return retval


def get_suffix_list_file_name():
    # type: () -> Text
    '''Get the file name for the public-suffix list data file

:returns: The filename for the datafile in this module.
:rtype: ``str``'''
    # TODO: automatically update the suffix list data file
    # <https://publicsuffix.org/list/effective_tld_names.dat>
    retval = resource_filename('gs.dmarc', 'suffixlist.txt')
    return retval
